rollout/trafficrouting/gatewayapi/gatewayapi.go

```diff
@@ -0,0 +1,219 @@
+func NewDynamicClient(di dynamic.Interface, namespace string) dynamic.ResourceInterface {
+	return di.Resource(GetMappingGVR()).Namespace(namespace)
+}
+
+func GetMappingGVR() schema.GroupVersionResource {
+	return toMappingGVR(defaults.GetGatewayAPIGroupVersion())
+}
+
+func toMappingGVR(apiVersion string) schema.GroupVersionResource {
+	parts := strings.Split(apiVersion, "/")
+	group := parts[0]
+	resourcename := apiGroupToResource[group]
+	version := parts[len(parts)-1]
+	return schema.GroupVersionResource{
+		Group:    group,
+		Version:  version,
+		Resource: resourcename,
+	}
+}
+
+func NewReconciler(cfg *ReconcilerConfig) *Reconciler {
+	reconciler := Reconciler{
+		Rollout:  cfg.Rollout,
+		Client:   cfg.Client,
+		Recorder: cfg.Recorder,
+	}
+	return &reconciler
+}
+
+func (r *Reconciler) UpdateHash(canaryHash, stableHash string, additionalDestinations ...v1alpha1.WeightDestination) error {
+	return nil
+}
+
+func (r *Reconciler) SetWeight(desiredWeight int32, additionalDestinations ...v1alpha1.WeightDestination) error {
+	ctx := context.TODO()
+	rollout := r.Rollout
+	httpRouteName := rollout.Spec.Strategy.Canary.TrafficRouting.GatewayAPI.HTTPRoute
+	httpRoute, err := r.Client.Get(ctx, httpRouteName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	canaryServiceName := rollout.Spec.Strategy.Canary.CanaryService
+	stableServiceName := rollout.Spec.Strategy.Canary.StableService
+	rules, isFound, err := unstructured.NestedSlice(httpRoute.Object, "spec", "rules")
+	if err != nil {
+		return err
+	}
+	if !isFound {
+		return errors.New("spec.rules field was not found in httpRoute")
+	}
+	backendRefs, err := getBackendRefs(rules)
+	if err != nil {
+		return err
+	}
+	if backendRefs == nil {
+		return errors.New("spec.rules.backendRefs field was not found in httpRoute")
+	}
+	canaryService, err := getService(canaryServiceName, backendRefs)
+	if err != nil {
+		return err
+	}
+	if canaryService == nil {
+		return errors.New("canaryService was not found in httpRoute")
+	}
+	err = unstructured.SetNestedField(canaryService, int64(desiredWeight), "weight")
+	if err != nil {
+		return err
+	}
+	stableService, err := getService(stableServiceName, backendRefs)
+	if err != nil {
+		return err
+	}
+	if stableService == nil {
+		return errors.New("stableService was not found in httpRoute")
+	}
+	err = unstructured.SetNestedField(stableService, int64(100-desiredWeight), "weight")
+	if err != nil {
+		return err
+	}
+	rules, err = mergeBackendRefs(rules, backendRefs)
+	if err != nil {
+		return err
+	}
+	err = unstructured.SetNestedSlice(httpRoute.Object, rules, "spec", "rules")
+	if err != nil {
+		return err
+	}
+	_, err = r.Client.Update(ctx, httpRoute, metav1.UpdateOptions{})
+	if err != nil {
+		msg := fmt.Sprintf("Error updating Gateway API %q: %s", httpRoute.GetName(), err)
+		r.sendWarningEvent(GatewayAPIUpdateError, msg)
+	}
+	return err
+}
+
+func getService(serviceName string, services []interface{}) (map[string]interface{}, error) {
```

Comment:
nit: change the parameter name `services` to `rules`

Reply:
I think this actually makes more sense as backendRef which maps to https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.HTTPBackendRef if I am reading it correctly it is going through the list of backendRefs and finding the service that it goes to.